text/0000-rename.md

```diff
@@ -0,0 +1,91 @@
+- Feature Name: rename_attr
+- Start Date: 2015-06-03
+- RFC PR: (leave this empty)
+- Rust Issue: (leave this empty)
+
+# Summary
+
+Add a `#[renamed(from="old_name", since="...")]` attribute for renaming APIs in a forwards
+*and* backwards compatible way that requires no effort or trouble to downstream consumers.
+
+# Motivation
+
+Naming things is hard, and APIs churned enough near the release of 1.0 that we stabilized on
+some suboptimal names. We'll probably do it again in the future. Therefore it would be
+desirable to deprecate these in favour of a rename. However to avoid breaking everyone
+there needs to be some trick to make impls still work. There are a few ways to do this today,
+but all of them cause janky behaviour for *trait methods*, because they have a diverse.
+Set of implementors and default implementations. For instance consumers of one
+name might get different behaviour. However if we could alias functions/methods under new names
+in a first-class way, we could avoid any problems.
+
+# Detailed design
+
+Add an attribute for functions `#[renamed(from="old_name", since="...")]`. This should have the following behaviour:
+
+* Any reference to the current or old_name should resolve to the current
+* The old name can be silently shadowed, to avoid namespace pollution
+* The function can only be implemented under one name (deny duplicates)
+* Users of the old name should receive a `warn(renamed)`
+* Multiple renames should be able to co-exist (though this would be really unfortunate)
+
+In addition, the mythical rustfix could identify these attributes and automatically fix the
+source of anything using the old names.
+
+## An example:
+
+We decide that `size_hint` should *really* have been `len_hint`. We go to the Iterator trait and
+go from:
+
+```rust
+    /// Returns a lower and upper bound on the remaining length of the iterator.
+    ///
+    /// An upper bound of `None` means either there is no known upper bound, or
+    /// the upper bound does not fit within a `usize`.
+    #[inline]
+    #[stable(feature = "rust1", since = "1.0.0")]
+    fn size_hint(&self) -> (usize, Option<usize>) { (0, None) }
+```
+
+to
+
+```rust
+    /// Returns a lower and upper bound on the remaining length of the iterator.
+    ///
+    /// An upper bound of `None` means either there is no known upper bound, or
+    /// the upper bound does not fit within a `usize`.
+    #[inline]
+    #[stable(feature = "rust1", since = "1.0.0")]
+    #[renamed(from = "size_hint", since = "1.2.0")]
+    fn len_hint(&self) -> (usize, Option<usize>) { (0, None) }
+```
+
+Now everyone who implements or consumes `size_hint` receives a `warn(renamed)`.
+A trivial grep party (or ideally rustfix), updates everything to use `len_hint`,
+and the warnings go away.
+
+# Drawbacks
+
+There are none beyond the usual "adding stuff"
+
+# Alternatives
+
+Enable full `fn` aliases via a `fn foo = bar;` syntax. This would allow renaming
```

Comment:
Keep in mind this is an issue specifically for methods (the RFC could definitely be more clear about that).
Adding `use` syntax for methods could work, but the attribute method was proposed because it requires no syntax changes.